add MetricExtractor

rbroc opened the pull request:
Add extractor taking probability distribution as input and returning metrics of the distribution (e.g. entropy)

rbroc commented:
@tyarkoni @adelavega  I’ve started looking into the entropy/other-probability-metric extractor in this PR. First step is creating a new stimulus type, that would be the input for the `ProbDistMetricExtractor`. We talked about something like a `ProbStim` or a `DictStim`, but I’m not sure where either of the two would fit in the `Stim` taxonomy, since stimuli are organized by modality.

The way I was thinking about it would be to create a new type of stimuli which are more "numeric/algebraic" in type, something like `VectorStims` (or `MatrixStim`) where the data is numpy vectors/arrays and an optional label attribute stores labels for each vector (if that applies). Extractors like the `ProbDistMetricExtractor` (will come up with a better name) could apply to these types, but other stuff like extractors computing autoregressive features of time series could be another example. Probability distributions could either simply be a `VectorStim` or have a specific `ProbDistStim` subclass. 

Open to suggestions on whether that could work, or whether you'd rather do things differently.

rbroc commented:
@adelavega @tyarkoni This should now be ready for a first review. I've implemented new stimulus and extractor (the way I explained in previous comment), and added tests for both which pass locally.
Can't see my latest commits here, something weird must have happened while Github was down. But all commits are there in the branch of my forked repo this PR comes from.

adelavega commented on pliers/stimuli/vector.py:
```diff
@@ -0,0 +1,109 @@
+            that applies.
+        data_dict (dict): If defined, overwrites array and labels. Keys are 
+            passed to labels argument, values are passed to array argument.
+        filename (str): Path to file, if array has to be read from file. 
```
I wonder if this stim should have so many options for loading, or if should keep it simple and leave that to the user (they can load, and then create stim)

rbroc replied:
leaving read from and save to tsv option only (reads both from `filename` and `url`).

adelavega commented on pliers/stimuli/vector.py:
```diff
@@ -0,0 +1,109 @@
+    def data(self):
+        return self.array
+
+    def save(self, path, as_json=True):
```
I'd think tsv might be a better default?

adelavega commented on pliers/stimuli/vector.py:
```diff
@@ -0,0 +1,109 @@
+        order=None, name=None, sort_data=None, url=None):
+
+        if filename is not None:
+            if filename.endswith('.json'):
```
Sometimes filename endings are not sufficient to distinguish file type. Probably edge case though.

adelavega commented on pliers/stimuli/vector.py:
```diff
@@ -0,0 +1,109 @@
+                with open(filename) as f:
+                    data_dict = json.loads(f.read())
+            else:
+                df = pd.read_csv(filename, sep='\t')
```
I guess one reason I would prefer to keep this simple and let the user do the reading, is that you might want to load in a different way, and it's a hassle to support all the relevant kwargs. 

For example, reading a csv with commas as separators.

rbroc replied:
forcing to read from tsv.

rbroc commented on pliers/extractors/vector.py:
```diff
@@ -0,0 +1,67 @@
+                                 'functions')
+        for idx, f in enumerate(functions):
+            if type(f) == str:
+                f = f.replace('numpy', 'np')
```
It's simply because numpy is already imported as np here, so forcing to to np would avoid having to reimport with full package name specification.

adelavega replied:
Yes, that made sense to me. My point was that more generally `np` is not an "official" shortname for numpy, just a convention. So it might be safer and more genral to just generalize this to any package the user may have installed, and require the complete package name, etc..

rbroc replied:
okay, so that implies we would try to import whatever package the function comes from? Makes sense, if it is not a problem to potentially reimport numpy here without shortname.

rbroc replied:
read now you comment below, which makes things clear.

rbroc commented on pliers/extractors/vector.py:
```diff
@@ -0,0 +1,67 @@
+    _log_attributes = ('functions',)
+
+    def __init__(self, functions=['np.mean', 'np.std', 'np.max', 'np.min'], 
+                 var_names=None, **kwargs):
```
I guess the alternative would be asking the user to pass a dictionary of dictionaries with function names as keys and kwargs dictionary as values. 
Leaving it as it is for now, but open to alternative options :)

rbroc commented on pliers/extractors/vector.py:
```diff
@@ -0,0 +1,67 @@
+                f = f.replace('numpy', 'np')
+                try:
+                    f_list = f.split('.')
+                    if len(f_list) > 1 and 'lambda' not in f:
```
now importing via `import_module`. Also removed option to evaluate lambda (user can define function then pass it directly, I guess).

rbroc commented:
done implementing @adelavega's suggestions (thank you!) and ready for review again. Tests seem to pass if you look at the travis log, but coverall times out for some weird reason.

rbroc commented:
@tyarkoni @adelavega I've rebased this to [Tal's PR](https://github.com/tyarkoni/pliers/pull/394) and:
- Removed `VectorStim` and related tests
- Moved `MetricExtractor` to new `misc` type and set `SeriesStim` as input
- Added option to subset `SeriesStim.data` in input to `MetricExtractor` by passing list of Series indices as argument (added test for this)
- Fixed a little bug in `BertLM`. Tokens are capitalized before being used as column names (to avoid overlap with `onset,` `duration` and other metadata columns). But this caused distinct one-character non-ascii tokens to be "capitalized" into the same character, and therefore collapsed into same column when calling `to_df. Now tokens containing non-ascii are left unchanged.
- Added test to make sure implicit conversion works when feeding `ExtractorResult` as input to `MetricExtractor`.
If we're not planning on creating a `DataFrameStim` within this PR, this is ready for review.

adelavega commented on pliers/extractors/misc.py:
```diff
@@ -0,0 +1,78 @@
+            (e.g. 'numpy.mean'). Custom functions returning integers or iterables 
+            can also be passed by passing the function itself.
+        var_names (list): optional list of custom alias names for each metric
+        subset_idx (list): subset of Series indices to compute metric on.
```
Could this be label based? On Tal's PR I commented that labels are not a requirement for pandas Series, but by default they do get added when using the converter. It seems quite clunky to have to create a boolean index and then pass it to this extractor, rather than labels. 

If we're going down this route, we may even want to allow for regex to filter columns. Although I think in most cases we'l want to restrict columns at the level of the first extractor's output.

rbroc replied:
not sure I understand. The labels are added as index at conversion, right?

rbroc replied:
right, this was just a terminology issue, should be solved by updated docstring.

adelavega commented on pliers/extractors/misc.py:
```diff
@@ -0,0 +1,78 @@
+        for idx, f in enumerate(functions):
+            if isinstance(f, str):
+                try:
+                    f_mod = f.split('.')[:-1]
```
```suggestion
                    f_mod, f_func = f.rsplit('.', 1) 
```

adelavega commented on pliers/extractors/misc.py:
```diff
@@ -0,0 +1,78 @@
+                try:
+                    f_mod = f.split('.')[:-1]
+                    f_func = f.split('.')[-1]
+                    functions[idx] = getattr(import_module('.'.join(f_mod)),
```
```suggestion
                    functions[idx] = getattr(import_module(f_mod),
```

rbroc commented:
implemented @adelavega's suggestions.

tyarkoni commented on pliers/extractors/misc.py:
```diff
@@ -0,0 +1,75 @@
+            series = stim.data
+        for f in self.functions:
+            metrics = f(series, **self.kwargs)
+            if isinstance(metrics, Iterable):
```
One issue here is that strings are `Iterable`, so if a function returns a string, it'll be turned into a list of characters. You can use `pliers.utils.isiterable()`, which will only check for certain types. I don't think it catches pandas series though, so we may want to add that to the type tuple in `isiterable`.

rbroc replied:
added `pd.Series` to `isiterable` (noticed that it does not include numpy arrays, btw) and now using that to check for `type` of `metrics`.

rbroc commented:
implemented Tal's suggestions, @tyarkoni @adelavega this should be ready to merge.
